docs/use_cases/retrieval_augmented_generation.md

```diff
@@ -1,15 +1,236 @@
+
+```
+
+Great. Next we will transform the PDF into texts and store them as vector embeddings in a basic vector store. We then perform a simple similarity search on our documents (aka retrieval).
+
+```python
+
+from langchain.vectorstores import FAISS
+
+from langchain.embeddings import HuggingFaceEmbeddings
+
+from langchain.document_loaders import PyPDFLoader
+
+# Loading the PDF as text and splitting it into pages
+
+loader = PyPDFLoader("tesla_q2_2023.pdf")
+
+pages = loader.load_and_split()
+
+# Creating indexed vector embeddings from our documents
+
+vectorstore = FAISS.from_documents(pages, HuggingFaceEmbeddings())
+
+# Searching for vectors similar to our input query with KNN
+
+docs = vectorstore.similarity_search("What was the revenue?", k=2)
+
+# Printing the documents to see what was retrieved
+
+docs
+
+```
+
+I will only show the relevant part of the documents here for aesthetic reasons, so keep in mind that the following is *not* the full output of docs:
+
+```console
+
+Revenue Total revenue grew 47% YoY in Q2 to $24.9B. 
+
+YoY, revenue was impacted by the following items: 
+
++growth in vehicle deliveries 
+
++growth in other parts of the business 
+
+-reduced ASP YoY (excluding FX impact) 
+
+-negative FX impact of $0.6B1
+
+```
+
+Let's create our context in the simplest way possible by attaching the results to a single string:
+
+```python
+
+context = ""
+
+for doc in docs:
+
+	context += doc.page_content
+
+```
+
+Finally, we will use the same tokenization and generation pipeline again as we did earlier, just this time we add the context we retrieved in front of our question.
+
+```python
+
+# Tokenizing our input string
+
+inputs = tokenizer(f'''{context} \n\nWhat was the revenue of Tesla for Q2 2023? \nAnswer:''',
+
+return_tensors="pt",
+
+return_attention_mask=False
+
+)
+
+# Generating a response based on our context with a max length to 1000 characters
+
+outputs = model.generate(**inputs, max_length=1000)
+
+text = tokenizer.batch_decode(outputs)[0]
+
+# Splitting the output by "\n\n" to only print the question and answer part
+
+print(text.split("\n\n")[1])
+
+```
+
+```console
+
+What was the revenue of Tesla for Q2 2023? 
+
+Answer: Tesla's revenue for Q2 2023 was $24.9 billion.
+
+```
+
+Et voilà! If you go back to the passage from the original document, this is the correct answer. More than 10x higher than what our model hallucinated earlier - RAG saved us from what could've lead to a terrible decision. 
+
+### Conclusion
+
+I hope that I could make clear how retrieval can add value to generation. It's not a silver bullet, but it has a lot of advantages and few disadvantages. Only in rare cases does retrieving relevant context degrade the quality of a model's response. Put simply: you have to make sure that the documents you're using do *actually* include the answers you're looking for. Irrelevant context can hurt the generative process. As with all things Data Science: garbage in, garbage out.
```

Comment:
1. I removed the degradation part completely since it's not that important for the article. I only left the final statement to remind people that RAG isn't magic and can only work with the data one provides it with.

2. I slightly changed "...documents you're retrieving..." to "...documents you're retrieving from..." since the sentence is about the ground truth documents from the external database.